leftwm-layouts/src/geometry/direction.rs

```diff
@@ -0,0 +1,295 @@
+
+    // We are all the way up, no neighbor available
+    if current_rect.y <= 0 {
+        return None;
+    }
+
+    let right_points = (current_rect.x + 1..current_rect.x + current_rect.w as i32 - 1).collect();
+    let up_points = (0..current_rect.y).rev().collect();
+
+    search_nearest_neighbor(&right_point, &up_points, rects, false)
+}
+
+// Find the east neighbor starting from a given `Rect` with index `current` in an array of
+// [`Rect`].
+fn find_east(rects: &[Rect], current: usize, display_width: u32) -> Option<usize> {
+    let Some(current_rect) = rects.get(current).or(None) else { return None };
+
+    // We are all the way right, no neighbor available
+    if current_rect.x + current_rect.w as i32 >= display_width as i32 {
+        return None;
+    }
+
+    let down_points = (current_rect.y + 1..current_rect.y + current_rect.h as i32).collect();
+    let right_points = (current_rect.x + current_rect.w as i32 + 1..=display_width as i32).collect();
+
+    search_nearest_neighbor(&down_points, &right_points, rects, true)
+}
+
+// Find the south neighbor starting from a given `Rect` with index `current` in an array of
+// [`Rect`].
+fn find_south(rects: &[Rect], current: usize, display_height: u32) -> Option<usize> {
+    let Some(current_rect) = rects.get(current).or(None) else { return None };
+
+    // We are at the bottom, no neighbor available
+    if current_rect.y + current_rect.h as i32 >= display_height as i32 {
+        return None;
+    }
+
+    let right_points = (current_rect.x + 1..current_rect.x + current_rect.w as i32).collect();
+    let down_points =
+        (current_rect.y + current_rect.h as i32 + 1..=display_height as i32).collect();
+
+    search_nearest_neighbor(&right_points, &down_points, rects, false)
+}
+
+// Find the west neighbor starting from a given `Rect` with index `current` in an array of
+// [`Rect`].
+fn find_west(rects: &[Rect], current: usize) -> Option<usize> {
+    let Some(current_rect) = rects.get(current).or(None) else { return None };
+
+    // We are all the way left; no neighbor available
+    if current_rect.x <= 0 {
+        return None;
+    }
+
+    let left_points = (0..=current_rect.x + -1).rev().collect();
+    let down_points = (current_rect.y + 1..current_rect.y + current_rect.h as i32).collect();
+
+    search_nearest_neighbor(&left_points, &down_points, rects, false)
+}
+
+fn search_nearest_neighbor(
+    first_direction: &Vec<i32>,
+    second_direction: &Vec<i32>,
+    rects: &[Rect],
+    inverted: bool,
+) -> Option<usize> {
+    for f in first_direction {
+        for s in second_direction {
+            for (i, r) in rects.iter().enumerate() {
+                if inverted {
+                    if r.contains((*s, *f)) {
+                        return Some(i);
+                    }
+                } else if r.contains((*f, *s)) {
+                    return Some(i);
+                }
+            }
+        }
+    }
+    None
+}
```

Comment:
A visual example how you could exclude rects from being considered. Of course if more than one rects is inside the zone, you'd need to figure out how to determine which rect is the most fitting. I'm sure you'd figure out a way to solve this :)

![Notes_230718_130710.jpg](https://github.com/leftwm/leftwm-layouts/assets/3533648/ad83077c-28fa-4c28-bb1a-46ace9016695)

Reply:
Correction: for **B** it should say `Can be excluded because their left edge is further right than the ref left edge`

Reply:
Hi @hertg thanks for your time. This is a great and easy idea. There are some corner cases that we can handle as follow, if you agree with me:
1. take the rect that is higher than the others if multiple rects can be chosen;
2. take the rect that is the left most if multiple rects can be chosen;
3. take the nearest rect if one rect is contained in another;

![2023-07-19-Note-19-20](https://github.com/leftwm/leftwm-layouts/assets/32301730/fdfaf6be-e0be-4c09-9ac1-25f338d31c0c)

My only concern is what to do if two rects overlap? E.g. two floating windows one on the other. Should we first take the newest one and then the older one ? How can we figure out which is the oldest one ? Maybe we can use the position in the Rects array ?

I will dig a bit into the coding part this weekend :smile:

Reply:
I had something as follows in mind, staying with the example of moving left/east:

1. Filter out all the rects that are not within the zone (as I decribed above)
2. Find the rect whose *right edge* is closest to the *left edge* of the `ref`
    
    > **Note**: If the rect's right edge goes beyond the *left edge* of the `ref` (overlapping), it shall be counted as if its *right edge* was actually exactly at the *left edge* of the `ref`.
    >
    > **Reasoning**: Because what is right of the *left edge* of `ref` is completely irrelevant when moving left.
    This also solves the issue of having to figure out which rect is "closer" if multiple were to overlap; they are both equally close.
    >
    > ![image](https://github.com/leftwm/leftwm-layouts/assets/3533648/8a9be463-cd41-4613-babb-a07ee577aa21)
3. If more than one rect are equally close, prefer the one that is higher up (for left/right movement) or most to the left (for up/down movement)


I'm sure there are some more edge cases that this logic doesn't address. Would be great to get those in form of a small sketch, I think visualizing helps a lot in understanding and finding solutions for such issues :)

Reply:
btw. you are welcome to introduce some helper methods on `Rect` to ease implementation and code readability. For example:

```rust
// this code may not be completely correct, just writing it out freehand without much thought

pub fn top_left_corner(&self) -> (i32, i32) {
  (self.x, self.y) // arguably not super helpful, may aid readability
}

pub fn top_right_corner(&self) -> (i32, i32) {
  (self.x + self.w, self.y)
}

pub fn bottom_right_corner(&self) -> (i32, i32) {
  (self.x + self.w, self.y + self.h)
}

pub fn bottom_left_corner(&self) -> (i32, i32) {
  (self.x, self.y + self.h)
}

pub fn top_edge(&self) -> i32 {
  self.x // arguably not super helpful, may aid readability
}

pub fn right_edge(&self) -> i32 {
  self.x + self.w
}

pub fn bottom_edge(&self) -> i32 {
  self.y + self.h
}

pub fn left_edge(&self) -> i32 {
  self.y // arguably not super helpful, may aid readability
}
```

Reply:
This appears to have been implemented so going to mark it resolved; please let me know if there are any concerns @hertg